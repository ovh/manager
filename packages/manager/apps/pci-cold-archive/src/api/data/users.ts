import { v6 } from '@ovh-ux/manager-core-api';
import { OBJECT_STORAGE_USER_ROLE } from '@/constants';

export type TS3Credentials = {
  userId: string;
  tenantId: string;
  access: string;
  secret: string;
};

export type TUser = {
  id: number;
  username: string;
  creationDate: string;
  description: string;
  openstackId: string;
  status: 'creating' | 'ok' | 'deleting' | 'deleted';
  access?: string;
  search?: string;
  password: string;
  s3Credentials?: TS3Credentials;
  roles: {
    id: string;
    name: string;
    description: string;
    permissions: string[];
  }[];
};

export const getUser = async (
  projectId: string,
  userId: number,
): Promise<TUser> => {
  const { data } = await v6.get<TUser>(
    `/cloud/project/${projectId}/user/${userId}`,
  );
  return data;
};

export const getAllUsers = async (projectId: string): Promise<TUser[]> => {
  const { data } = await v6.get<TUser[]>(`/cloud/project/${projectId}/user`);
  return data;
};

export const getS3Credentials = async (
  projectId: string,
  userId: string,
): Promise<TS3Credentials[]> => {
  const { data } = await v6.get<TS3Credentials[]>(
    `/cloud/project/${projectId}/user/${userId}/s3Credentials`,
  );
  return data;
};

export const deleteUser = async (
  projectId: string,
  userId: number,
  accessKey: string,
) => {
  await v6.delete(
    `/cloud/project/${projectId}/user/${userId}/s3Credentials/${accessKey}`,
  );
};

export const getUserStoragePolicy = async (
  projectId: string,
  userId: number,
): Promise<{ policy: string }> => {
  const { data } = await v6.get(
    `/cloud/project/${projectId}/user/${userId}/policy`,
  );
  return data;
};

export const postS3Secret = async (
  projectId: string,
  userId: number,
  userAccess: string,
): Promise<{ secret: string }> => {
  const { data } = await v6.post(
    `/cloud/project/${projectId}/user/${userId}/s3Credentials/${userAccess}/secret`,
  );

  return data;
};

export const importUserPolicy = async (
  projectId: string,
  userId: string,
  policy: string,
) => {
  const { data } = await v6.post(
    `/cloud/project/${projectId}/user/${userId}/policy`,
    {
      policy,
    },
  );
  return data;
};

export const generateS3Credentials = async (
  projectId: string,
  userId: number,
): Promise<TS3Credentials> => {
  const { data } = await v6.post<TS3Credentials>(
    `/cloud/project/${projectId}/user/${userId}/s3Credentials`,
  );
  return data;
};

export const createUser = async (
  projectId: string,
  description: string,
  role: string = OBJECT_STORAGE_USER_ROLE,
): Promise<TUser> => {
  const { data } = await v6.post<TUser>(`/cloud/project/${projectId}/user`, {
    description,
    role,
  });

  return data;
};
