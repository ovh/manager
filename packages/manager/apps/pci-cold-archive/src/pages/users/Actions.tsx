import { ApiError } from '@ovh-ux/manager-core-api';
import {
  ActionMenu,
  ActionMenuItem,
  useNotifications as useMRCNotifications,
} from '@ovh-ux/manager-react-components';

import { ShellContext } from '@ovh-ux/manager-react-shell-client';
import { saveAs } from 'file-saver';
import { useContext } from 'react';
import { Translation, useTranslation } from 'react-i18next';
import { useNavigate, useParams } from 'react-router-dom';
import { getUserStoragePolicy, TUser } from '@/api/data/users';
import { usePostS3Secret } from '@/api/hooks/useUsers';
import { DOWNLOAD_FILENAME, DOWNLOAD_TYPE } from '@/constants';
import { useNotifications } from '@/hooks/useNotifications';
import { COLD_ARCHIVE_TRACKING } from '@/tracking.constants';

export default function ActionsComponent({ user }: Readonly<{ user: TUser }>) {
  const { t } = useTranslation(['users', 'containers']);

  const { projectId } = useParams();

  const { addInfo } = useMRCNotifications();
  const { addSuccessMessage, addErrorMessage } = useNotifications({
    ns: 'users',
  });

  const navigate = useNavigate();

  const { tracking } = useContext(ShellContext).shell;

  const trackS3UsersPage = (action: string) => {
    tracking?.trackPage({
      name: `${COLD_ARCHIVE_TRACKING.PAGE_PREFIX}::${COLD_ARCHIVE_TRACKING.USER.MAIN}::${action}`,
      action: 'navigation',
    });
  };

  const trackS3UsersClick = (action: string) => {
    tracking?.trackClick({
      name: `${COLD_ARCHIVE_TRACKING.PAGE_PREFIX}::${COLD_ARCHIVE_TRACKING.USER.MAIN}::${action}`,
      type: 'action',
    });
  };

  const { postS3Secret: showSecretKey } = usePostS3Secret({
    projectId,
    userId: user.id,
    userAccess: user.access,
    onSuccess: ({ secret }) => {
      addInfo(
        <Translation ns="users">
          {(_t) => (
            <span
              dangerouslySetInnerHTML={{
                __html: _t(
                  'pci_projects_project_storages_containers_users_show_secret_key_success',
                  {
                    user: `<strong>${user.username}</strong>`,
                    secret: `<span style="color: var(--ods-color-critical-600); background: var(--ods-color-critical-050)">${secret}</span>`,
                  },
                ),
              }}
            />
          )}
        </Translation>,
        true,
      );
    },
    onError: (error: ApiError) => {
      addErrorMessage({
        i18nKey:
          'pci_projects_project_storages_containers_users_show_secret_key_error',
        values: {
          user: `<strong>${user.username}</strong>`,
        },
        error,
      });
    },
  });

  const downloadUserPolicyJson = () => {
    trackS3UsersClick(COLD_ARCHIVE_TRACKING.USER.ACTIONS.DOWNLOAD_POLICY);
    getUserStoragePolicy(projectId, user?.id)
      .then(({ policy }) => {
        trackS3UsersPage(
          `${COLD_ARCHIVE_TRACKING.USER.ACTIONS.DOWNLOAD_POLICY}_${COLD_ARCHIVE_TRACKING.STATUS.SUCCESS}`,
        );
        saveAs(new Blob([policy], { type: DOWNLOAD_TYPE }), DOWNLOAD_FILENAME);

        addSuccessMessage({
          i18nKey:
            'pci_projects_project_storages_containers_users_import_success_message',
          values: { user: user.username },
        });
      })
      .catch((error: ApiError) => {
        trackS3UsersPage(
          `${COLD_ARCHIVE_TRACKING.USER.ACTIONS.DOWNLOAD_POLICY}_${COLD_ARCHIVE_TRACKING.STATUS.ERROR}`,
        );

        addErrorMessage({
          i18nKey:
            'pci_projects_project_storages_containers_users_import_error_message',
          values: {
            user: `<strong>${user.username}</strong>`,
          },
          error,
        });
      });
  };

  const items: ActionMenuItem[] = [
    {
      id: 0,
      label: t('pci_projects_project_storages_containers_users_import_json'),
      onClick: () => {
        trackS3UsersClick(COLD_ARCHIVE_TRACKING.USER.ACTIONS.IMPORT_POLICY);
        navigate(`./import-policy?userId=${user.id}`);
      },
    },
    {
      id: 1,
      label: t('pci_projects_project_storages_containers_users_download_json'),
      onClick: downloadUserPolicyJson,
    },
    {
      id: 2,
      label: t(
        'pci_projects_project_storages_containers_users_download_rclone_file',
      ),
      onClick: () => {
        trackS3UsersClick(COLD_ARCHIVE_TRACKING.USER.ACTIONS.DOWNLOAD_RCLONE);
        navigate(`./rclone/download?userId=${user.id}`);
      },
    },
    {
      id: 3,
      label: t('pci_projects_project_storages_containers_users_see_secret_key'),
      onClick: () => {
        trackS3UsersClick(COLD_ARCHIVE_TRACKING.USER.ACTIONS.DISPLAY_SECRET);
        window.scrollTo(0, 0);
        showSecretKey();
      },
    },
    {
      id: 4,
      label: t(
        'containers:pci_projects_project_storages_containers_delete_label',
      ),
      onClick: () => {
        trackS3UsersClick(COLD_ARCHIVE_TRACKING.USER.ACTIONS.DELETE_POLICY);
        navigate(`./${user.id}/delete`);
      },
    },
  ];

  return <ActionMenu id={`${user.id}`} items={items} isCompact />;
}
