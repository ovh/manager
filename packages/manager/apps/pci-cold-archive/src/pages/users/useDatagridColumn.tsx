import { DataGridTextCell } from '@ovh-ux/manager-react-components';
import { format } from 'date-fns';
import { useTranslation } from 'react-i18next';
import { FilterCategories } from '@ovh-ux/manager-core-api';
import { TUser } from '@/api/data/users';
import ActionsComponent from './Actions';

export const useDatagridColumn = () => {
  const { t } = useTranslation('users');

  const columns = [
    {
      id: 'username',
      cell: (props: TUser) => (
        <DataGridTextCell>{props.username}</DataGridTextCell>
      ),
      label: t('pci_projects_project_storages_containers_users_username'),
      comparators: FilterCategories.String,
      isFilterable: true,
    },
    {
      id: 'creationDate',
      cell: (props: TUser) => (
        <DataGridTextCell>
          {format(props.creationDate, 'dd/MM/yyyy HH:mm')}
        </DataGridTextCell>
      ),
      label: t('pci_projects_project_storages_containers_users_creation_date'),
      comparators: FilterCategories.Date,
      isFilterable: true,
    },
    {
      id: 'description',
      cell: (props: TUser) => (
        <DataGridTextCell>{props.description}</DataGridTextCell>
      ),
      label: t('pci_projects_project_storages_containers_users_description'),
      comparators: FilterCategories.String,
      isFilterable: true,
    },
    {
      id: 'access',
      cell: (props: TUser) => (
        <DataGridTextCell>{props.access}</DataGridTextCell>
      ),
      label: t('pci_projects_project_storages_containers_users_accesskey'),
      comparators: FilterCategories.String,
      isFilterable: true,
    },
    {
      id: 'actions',
      cell: (props: TUser) => <ActionsComponent user={props} />,
      label: '',
      isSortable: false,
      isFilterable: false,
    },
  ];

  return columns;
};
