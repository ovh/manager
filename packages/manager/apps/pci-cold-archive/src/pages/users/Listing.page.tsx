import { FilterComparator } from '@ovh-ux/manager-core-api';
import {
  Datagrid,
  FilterAdd,
  FilterList,
  Notifications,
  useColumnFilters,
  useDataGrid,
  useNotifications,
} from '@ovh-ux/manager-react-components';
import { ColumnFilter } from '@ovh-ux/manager-react-components/dist/types/src/components/filters/filter-add.component';
import {
  OdsButton,
  OdsInput,
  OdsPopover,
  OdsText,
} from '@ovhcloud/ods-components/react';
import { useState } from 'react';
import { useTranslation } from 'react-i18next';
import { Outlet, useNavigate, useParams } from 'react-router-dom';
import { usePaginatedUsers } from '@/api/hooks/useUsers';
import { useDatagridColumn } from './useDatagridColumn';

export default function UsersListing() {
  const { t } = useTranslation('users');

  const { projectId } = useParams();
  const { pagination, setPagination, sorting, setSorting } = useDataGrid({
    desc: false,
    id: 'username',
  });

  const navigate = useNavigate();
  const { clearNotifications } = useNotifications();
  const { filters, addFilter, removeFilter } = useColumnFilters();
  const [searchField, setSearchField] = useState('');

  const columns = useDatagridColumn();

  const { paginatedUsers, refetch, isLoading } = usePaginatedUsers(
    projectId,
    pagination,
    sorting,
    filters,
  );

  const handleSearch = () => {
    setPagination({
      pageIndex: 0,
      pageSize: pagination.pageSize,
    });
    addFilter({
      key: 'username',
      value: searchField.trim(),
      comparator: FilterComparator.Includes,
      label: '',
    });
    setSearchField('');
  };

  return (
    <div className="flex flex-col gap-8">
      <div>
        <Notifications />
      </div>

      <OdsText preset="heading-4">
        {t('pci_projects_project_storages_containers_users_title')}
      </OdsText>

      <OdsText preset="paragraph">
        {t('pci_projects_project_storages_containers_users_user_description')}
      </OdsText>

      <div className="sm:flex items-center justify-between">
        <OdsButton
          label={t('pci_projects_project_storages_containers_users_add_user')}
          size="sm"
          icon="plus"
          color="primary"
          className="xs:mb-0.5 sm:mb-0"
          onClick={() => {
            clearNotifications();
            navigate('./new');
          }}
        />

        <div className="flex justify-center gap-4">
          <OdsButton
            size="sm"
            variant="outline"
            color="primary"
            className="xs:mb-0.5 sm:mb-0"
            onClick={refetch}
            icon="refresh"
            label=""
          />
          <OdsInput
            name="searchField"
            className="min-w-[15rem]"
            value={searchField}
            onOdsChange={({ detail }) => setSearchField(detail.value as string)}
            onKeyDown={(event) => {
              if (event.key === 'Enter') handleSearch();
            }}
          />
          <OdsButton
            label=""
            icon="magnifying-glass"
            size="sm"
            onClick={handleSearch}
          />

          <OdsButton
            slot="popover-trigger"
            id="popover-filter"
            size="sm"
            color="primary"
            label={t('pci-common:common_criteria_adder_filter_label')}
            variant="outline"
            icon="filter"
          />
          <OdsPopover triggerId="popover-filter">
            <FilterAdd
              columns={
                [...columns].filter(
                  (column) => column.isFilterable,
                ) as ColumnFilter[]
              }
              onAddFilter={(addedFilter, column) => {
                setPagination({
                  pageIndex: 0,
                  pageSize: pagination.pageSize,
                });
                addFilter({
                  ...addedFilter,
                  label: column.label,
                });
              }}
            />
          </OdsPopover>
        </div>
      </div>

      <FilterList filters={filters} onRemoveFilter={removeFilter} />

      <Datagrid
        columns={columns}
        items={isLoading ? [] : paginatedUsers?.rows}
        totalItems={paginatedUsers?.totalRows || 0}
        pagination={pagination}
        onPaginationChange={setPagination}
        sorting={sorting}
        onSortChange={setSorting}
        isLoading={isLoading}
      />

      <Outlet />
    </div>
  );
}
