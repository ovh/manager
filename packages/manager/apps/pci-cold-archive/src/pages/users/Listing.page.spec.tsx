import { render } from '@testing-library/react';
import React from 'react';
import { useNavigate } from 'react-router-dom';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import { useNotifications } from '@/hooks/useNotifications';
import { usePaginatedUsers } from '@/api/hooks/useUsers';
import { TUser } from '@/api/data/users';
import ListingPage from './Listing.page';
import { useTracking } from '@/hooks/useTracking';

vi.mock('@/pages/users/Actions.tsx');
vi.mock('date-fns', () => ({
  format: (date) => date,
}));

vi.mock('@/hooks/useTracking', () => ({
  useTracking: vi.fn(),
}));

vi.mock('@/hooks/useNotifications', () => ({
  useNotifications: vi.fn(),
}));

vi.mock('@/api/hooks/useUsers', () => ({
  usePaginatedUsers: vi.fn(),
}));

vi.mock('@ovh-ux/manager-react-shell-client', () => ({
  ShellContext: React.createContext({
    environment: {
      getUser: () => ({
        ovhSubsidiary: 'FR',
      }),
    },
  }),
}));

describe('ListingPage', () => {
  const mockNavigate = vi.fn();

  const mockAddSuccessMessage = vi.fn();
  const mockAddErrorMessage = vi.fn();

  const mockTrackConfirmAction = vi.fn();
  const mockTrackCancelAction = vi.fn();
  const mockTrackSuccessPage = vi.fn();
  const mockTrackErrorPage = vi.fn();
  const mockTrackActionClick = vi.fn();
  const mockTrackNavigationClick = vi.fn();
  const mockTackCopyClipboardClick = vi.fn();

  const mockPaginatedUsers = {
    rows: [
      {
        id: 1,
        creationDate: '2023-02-02',
      },
      {
        id: 2,
        creationDate: '2023-02-02',
      },
    ] as TUser[],
    totalRows: 2,
    pageCount: 0,
  };

  beforeEach(() => {
    vi.clearAllMocks();

    vi.mocked(useNavigate).mockReturnValue(mockNavigate);

    vi.mocked(useNotifications).mockReturnValue({
      addSuccessMessage: mockAddSuccessMessage,
      addErrorMessage: mockAddErrorMessage,
    });

    vi.mocked(useTracking).mockReturnValue({
      trackConfirmAction: mockTrackConfirmAction,
      trackCancelAction: mockTrackCancelAction,
      trackSuccessPage: mockTrackSuccessPage,
      trackErrorPage: mockTrackErrorPage,
      trackActionClick: mockTrackActionClick,
      trackNavigationClick: mockTrackNavigationClick,
      trackCopyClipboardClick: mockTackCopyClipboardClick,
    });
  });

  it('should match snapshot when is loading false', async () => {
    vi.mocked(usePaginatedUsers).mockReturnValue({
      paginatedUsers: mockPaginatedUsers,
      isPending: false,
    } as ReturnType<typeof usePaginatedUsers>);

    const { container } = render(<ListingPage />);

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot when is loading true', async () => {
    vi.mocked(usePaginatedUsers).mockReturnValue({
      paginatedUsers: mockPaginatedUsers,
      isPending: true,
    } as ReturnType<typeof usePaginatedUsers>);

    const { container } = render(<ListingPage />);

    expect(container).toMatchSnapshot();
  });
});
