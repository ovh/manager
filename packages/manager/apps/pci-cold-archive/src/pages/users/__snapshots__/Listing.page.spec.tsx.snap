// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ListingPage > should match snapshot when is loading false 1`] = `
<div>
  <div
    class="flex flex-col gap-8"
  >
    <div>
      <div
        data-testid="notifications"
      />
    </div>
    <ods-text
      preset="heading-4"
    >
      pci_projects_project_storages_containers_users_title
    </ods-text>
    <ods-text
      preset="paragraph"
    >
      pci_projects_project_storages_containers_users_user_description
    </ods-text>
    <div
      class="sm:flex items-center justify-between"
    >
      <ods-button
        class="xs:mb-0.5 sm:mb-0"
        color="primary"
        icon="plus"
        label="pci_projects_project_storages_containers_users_add_user"
        size="sm"
      />
      <div
        class="flex justify-center gap-4"
      >
        <ods-button
          class="xs:mb-0.5 sm:mb-0"
          color="primary"
          icon="refresh"
          label=""
          size="sm"
          variant="outline"
        />
        <ods-input
          class="min-w-[15rem]"
          name="searchField"
          value=""
        />
        <ods-button
          icon="magnifying-glass"
          label=""
          size="sm"
        />
        <ods-button
          color="primary"
          icon="filter"
          id="popover-filter"
          label="pci-common:common_criteria_adder_filter_label"
          size="sm"
          slot="popover-trigger"
          variant="outline"
        />
        <ods-popover
          trigger-id="popover-filter"
        >
          <div>
            <ods-form-field
              class="w-full"
            >
              <div
                slot="label"
              >
                <span
                  class="text-[--ods-color-heading] leading-[22px]"
                >
                  common_criteria_adder_column_label
                </span>
              </div>
              <ods-select
                data-testid="add-filter_select_idColumn"
                name="add-filter_select_idColumn"
                value="username"
              >
                <option
                  value="username"
                >
                  pci_projects_project_storages_containers_users_username
                </option>
                <option
                  value="creationDate"
                >
                  pci_projects_project_storages_containers_users_creation_date
                </option>
                <option
                  value="description"
                >
                  pci_projects_project_storages_containers_users_description
                </option>
                <option
                  value="access"
                >
                  pci_projects_project_storages_containers_users_accesskey
                </option>
              </ods-select>
            </ods-form-field>
          </div>
          <div>
            <ods-form-field
              class="mt-2 w-full"
            >
              <div
                slot="label"
              >
                <span
                  class="text-[--ods-color-heading] leading-[22px]"
                >
                  common_criteria_adder_operator_label
                </span>
              </div>
              <div>
                <ods-select
                  name="add-operator-username"
                  value="includes"
                >
                  <option
                    value="includes"
                  >
                    common_criteria_adder_operator_includes
                  </option>
                  <option
                    value="starts_with"
                  >
                    common_criteria_adder_operator_starts_with
                  </option>
                  <option
                    value="ends_with"
                  >
                    common_criteria_adder_operator_ends_with
                  </option>
                  <option
                    value="is_equal"
                  >
                    common_criteria_adder_operator_is_equal
                  </option>
                  <option
                    value="is_different"
                  >
                    common_criteria_adder_operator_is_different
                  </option>
                </ods-select>
              </div>
            </ods-form-field>
          </div>
          <div>
            <ods-form-field
              class="mt-2 w-full"
            >
              <div
                slot="label"
              >
                <span
                  class="text-[--ods-color-heading] leading-[22px]"
                >
                  common_criteria_adder_value_label
                </span>
              </div>
              <ods-input
                class="border"
                data-testid="filter-add_value-input"
                name="filter-add_value-input"
                type="text"
                value=""
              />
            </ods-form-field>
          </div>
          <div>
            <ods-button
              class="mt-4 w-full filter-add-button-submit"
              data-testid="filter-add_submit"
              is-disabled="true"
              label="common_criteria_adder_submit_label"
              size="sm"
            />
          </div>
        </ods-popover>
      </div>
    </div>
    <div>
      <div
        class="contents px-[1px] "
      >
        <ods-table
          class="overflow-x-visible"
        >
          <!---->
          <table
            class="w-full border-collapse"
            style="--expander-column-width: 2.5rem;"
          >
            <thead>
              <tr>
                <th
                  class="text-left pl-4 h-11 whitespace-nowrap "
                >
                  <div
                    class="cursor-pointer select-none"
                    data-testid="header-username"
                  >
                    <span>
                      pci_projects_project_storages_containers_users_username
                    </span>
                    <span
                      class="align-middle inline-block h-4 -mt-6"
                    >
                      <ods-icon
                        name="arrow-up"
                      />
                    </span>
                  </div>
                </th>
                <th
                  class="text-left pl-4 h-11 whitespace-nowrap "
                >
                  <div
                    class="cursor-pointer select-none"
                    data-testid="header-creationDate"
                  >
                    <span>
                      pci_projects_project_storages_containers_users_creation_date
                    </span>
                    <span
                      class="align-middle inline-block h-4 -mt-6"
                    >
                      <ods-icon
                        class="invisible"
                        name="arrow-down"
                      />
                    </span>
                  </div>
                </th>
                <th
                  class="text-left pl-4 h-11 whitespace-nowrap "
                >
                  <div
                    class="cursor-pointer select-none"
                    data-testid="header-description"
                  >
                    <span>
                      pci_projects_project_storages_containers_users_description
                    </span>
                    <span
                      class="align-middle inline-block h-4 -mt-6"
                    >
                      <ods-icon
                        class="invisible"
                        name="arrow-down"
                      />
                    </span>
                  </div>
                </th>
                <th
                  class="text-left pl-4 h-11 whitespace-nowrap "
                >
                  <div
                    class="cursor-pointer select-none"
                    data-testid="header-access"
                  >
                    <span>
                      pci_projects_project_storages_containers_users_accesskey
                    </span>
                    <span
                      class="align-middle inline-block h-4 -mt-6"
                    >
                      <ods-icon
                        class="invisible"
                        name="arrow-down"
                      />
                    </span>
                  </div>
                </th>
                <th
                  class="text-left pl-4 h-11 whitespace-nowrap "
                >
                  <div
                    class=""
                    data-testid="header-actions"
                  >
                    <span />
                    <span
                      class="align-middle inline-block h-4 -mt-6"
                    >
                      <ods-icon
                        class="invisible"
                        name="arrow-down"
                      />
                    </span>
                  </div>
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                class="border-solid border-[1px] h-[3.25rem] border-[--ods-color-blue-200]"
              >
                <td
                  class="text-left pl-4"
                >
                  <ods-text
                    preset="span"
                  >
                    common_empty_text_cell
                  </ods-text>
                </td>
                <td
                  class="text-left pl-4"
                >
                  <ods-text
                    preset="span"
                  >
                    2023-02-02
                  </ods-text>
                </td>
                <td
                  class="text-left pl-4"
                >
                  <ods-text
                    preset="span"
                  >
                    common_empty_text_cell
                  </ods-text>
                </td>
                <td
                  class="text-left pl-4"
                >
                  <ods-text
                    preset="span"
                  >
                    common_empty_text_cell
                  </ods-text>
                </td>
                <td
                  class="text-left pl-4"
                />
              </tr>
              <tr
                class="border-solid border-[1px] h-[3.25rem] border-[--ods-color-blue-200]"
              >
                <td
                  class="text-left pl-4"
                >
                  <ods-text
                    preset="span"
                  >
                    common_empty_text_cell
                  </ods-text>
                </td>
                <td
                  class="text-left pl-4"
                >
                  <ods-text
                    preset="span"
                  >
                    2023-02-02
                  </ods-text>
                </td>
                <td
                  class="text-left pl-4"
                >
                  <ods-text
                    preset="span"
                  >
                    common_empty_text_cell
                  </ods-text>
                </td>
                <td
                  class="text-left pl-4"
                >
                  <ods-text
                    preset="span"
                  >
                    common_empty_text_cell
                  </ods-text>
                </td>
                <td
                  class="text-left pl-4"
                />
              </tr>
            </tbody>
          </table>
        </ods-table>
      </div>
      <ods-pagination
        class="flex xs:justify-start md:justify-end my-8"
        default-current-page="1"
        default-items-per-page="10"
        total-items="2"
        total-pages="1"
      >
        <span
          class="mr-3"
          slot="before-total-items"
        >
          common_pagination_of
        </span>
        <span
          class="ml-3"
          slot="after-total-items"
        >
          common_pagination_results
        </span>
      </ods-pagination>
    </div>
    Outlet
  </div>
</div>
`;

exports[`ListingPage > should match snapshot when is loading true 1`] = `
<div>
  <div
    class="flex flex-col gap-8"
  >
    <div>
      <div
        data-testid="notifications"
      />
    </div>
    <ods-text
      preset="heading-4"
    >
      pci_projects_project_storages_containers_users_title
    </ods-text>
    <ods-text
      preset="paragraph"
    >
      pci_projects_project_storages_containers_users_user_description
    </ods-text>
    <div
      class="sm:flex items-center justify-between"
    >
      <ods-button
        class="xs:mb-0.5 sm:mb-0"
        color="primary"
        icon="plus"
        label="pci_projects_project_storages_containers_users_add_user"
        size="sm"
      />
      <div
        class="flex justify-center gap-4"
      >
        <ods-button
          class="xs:mb-0.5 sm:mb-0"
          color="primary"
          icon="refresh"
          label=""
          size="sm"
          variant="outline"
        />
        <ods-input
          class="min-w-[15rem]"
          name="searchField"
          value=""
        />
        <ods-button
          icon="magnifying-glass"
          label=""
          size="sm"
        />
        <ods-button
          color="primary"
          icon="filter"
          id="popover-filter"
          label="pci-common:common_criteria_adder_filter_label"
          size="sm"
          slot="popover-trigger"
          variant="outline"
        />
        <ods-popover
          trigger-id="popover-filter"
        >
          <div>
            <ods-form-field
              class="w-full"
            >
              <div
                slot="label"
              >
                <span
                  class="text-[--ods-color-heading] leading-[22px]"
                >
                  common_criteria_adder_column_label
                </span>
              </div>
              <ods-select
                data-testid="add-filter_select_idColumn"
                name="add-filter_select_idColumn"
                value="username"
              >
                <option
                  value="username"
                >
                  pci_projects_project_storages_containers_users_username
                </option>
                <option
                  value="creationDate"
                >
                  pci_projects_project_storages_containers_users_creation_date
                </option>
                <option
                  value="description"
                >
                  pci_projects_project_storages_containers_users_description
                </option>
                <option
                  value="access"
                >
                  pci_projects_project_storages_containers_users_accesskey
                </option>
              </ods-select>
            </ods-form-field>
          </div>
          <div>
            <ods-form-field
              class="mt-2 w-full"
            >
              <div
                slot="label"
              >
                <span
                  class="text-[--ods-color-heading] leading-[22px]"
                >
                  common_criteria_adder_operator_label
                </span>
              </div>
              <div>
                <ods-select
                  name="add-operator-username"
                  value="includes"
                >
                  <option
                    value="includes"
                  >
                    common_criteria_adder_operator_includes
                  </option>
                  <option
                    value="starts_with"
                  >
                    common_criteria_adder_operator_starts_with
                  </option>
                  <option
                    value="ends_with"
                  >
                    common_criteria_adder_operator_ends_with
                  </option>
                  <option
                    value="is_equal"
                  >
                    common_criteria_adder_operator_is_equal
                  </option>
                  <option
                    value="is_different"
                  >
                    common_criteria_adder_operator_is_different
                  </option>
                </ods-select>
              </div>
            </ods-form-field>
          </div>
          <div>
            <ods-form-field
              class="mt-2 w-full"
            >
              <div
                slot="label"
              >
                <span
                  class="text-[--ods-color-heading] leading-[22px]"
                >
                  common_criteria_adder_value_label
                </span>
              </div>
              <ods-input
                class="border"
                data-testid="filter-add_value-input"
                name="filter-add_value-input"
                type="text"
                value=""
              />
            </ods-form-field>
          </div>
          <div>
            <ods-button
              class="mt-4 w-full filter-add-button-submit"
              data-testid="filter-add_submit"
              is-disabled="true"
              label="common_criteria_adder_submit_label"
              size="sm"
            />
          </div>
        </ods-popover>
      </div>
    </div>
    <div>
      <div
        class="contents px-[1px] "
      >
        <ods-table
          class="overflow-x-visible"
        >
          <!---->
          <table
            class="w-full border-collapse"
            style="--expander-column-width: 2.5rem;"
          >
            <thead>
              <tr>
                <th
                  class="text-left pl-4 h-11 whitespace-nowrap "
                >
                  <div
                    class="cursor-pointer select-none"
                    data-testid="header-username"
                  >
                    <span>
                      pci_projects_project_storages_containers_users_username
                    </span>
                    <span
                      class="align-middle inline-block h-4 -mt-6"
                    >
                      <ods-icon
                        name="arrow-up"
                      />
                    </span>
                  </div>
                </th>
                <th
                  class="text-left pl-4 h-11 whitespace-nowrap "
                >
                  <div
                    class="cursor-pointer select-none"
                    data-testid="header-creationDate"
                  >
                    <span>
                      pci_projects_project_storages_containers_users_creation_date
                    </span>
                    <span
                      class="align-middle inline-block h-4 -mt-6"
                    >
                      <ods-icon
                        class="invisible"
                        name="arrow-down"
                      />
                    </span>
                  </div>
                </th>
                <th
                  class="text-left pl-4 h-11 whitespace-nowrap "
                >
                  <div
                    class="cursor-pointer select-none"
                    data-testid="header-description"
                  >
                    <span>
                      pci_projects_project_storages_containers_users_description
                    </span>
                    <span
                      class="align-middle inline-block h-4 -mt-6"
                    >
                      <ods-icon
                        class="invisible"
                        name="arrow-down"
                      />
                    </span>
                  </div>
                </th>
                <th
                  class="text-left pl-4 h-11 whitespace-nowrap "
                >
                  <div
                    class="cursor-pointer select-none"
                    data-testid="header-access"
                  >
                    <span>
                      pci_projects_project_storages_containers_users_accesskey
                    </span>
                    <span
                      class="align-middle inline-block h-4 -mt-6"
                    >
                      <ods-icon
                        class="invisible"
                        name="arrow-down"
                      />
                    </span>
                  </div>
                </th>
                <th
                  class="text-left pl-4 h-11 whitespace-nowrap "
                >
                  <div
                    class=""
                    data-testid="header-actions"
                  >
                    <span />
                    <span
                      class="align-middle inline-block h-4 -mt-6"
                    >
                      <ods-icon
                        class="invisible"
                        name="arrow-down"
                      />
                    </span>
                  </div>
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                class="border-solid border-[1px] h-[3.25rem] border-[--ods-color-blue-200]"
              >
                <td
                  class="text-left pl-4"
                >
                  <ods-text
                    preset="span"
                  >
                    common_empty_text_cell
                  </ods-text>
                </td>
                <td
                  class="text-left pl-4"
                >
                  <ods-text
                    preset="span"
                  >
                    2023-02-02
                  </ods-text>
                </td>
                <td
                  class="text-left pl-4"
                >
                  <ods-text
                    preset="span"
                  >
                    common_empty_text_cell
                  </ods-text>
                </td>
                <td
                  class="text-left pl-4"
                >
                  <ods-text
                    preset="span"
                  >
                    common_empty_text_cell
                  </ods-text>
                </td>
                <td
                  class="text-left pl-4"
                />
              </tr>
              <tr
                class="border-solid border-[1px] h-[3.25rem] border-[--ods-color-blue-200]"
              >
                <td
                  class="text-left pl-4"
                >
                  <ods-text
                    preset="span"
                  >
                    common_empty_text_cell
                  </ods-text>
                </td>
                <td
                  class="text-left pl-4"
                >
                  <ods-text
                    preset="span"
                  >
                    2023-02-02
                  </ods-text>
                </td>
                <td
                  class="text-left pl-4"
                >
                  <ods-text
                    preset="span"
                  >
                    common_empty_text_cell
                  </ods-text>
                </td>
                <td
                  class="text-left pl-4"
                >
                  <ods-text
                    preset="span"
                  >
                    common_empty_text_cell
                  </ods-text>
                </td>
                <td
                  class="text-left pl-4"
                />
              </tr>
            </tbody>
          </table>
        </ods-table>
      </div>
      <ods-pagination
        class="flex xs:justify-start md:justify-end my-8"
        default-current-page="1"
        default-items-per-page="10"
        total-items="2"
        total-pages="1"
      >
        <span
          class="mr-3"
          slot="before-total-items"
        >
          common_pagination_of
        </span>
        <span
          class="ml-3"
          slot="after-total-items"
        >
          common_pagination_results
        </span>
      </ods-pagination>
    </div>
    Outlet
  </div>
</div>
`;
